devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,236 @@
+        else:
+            return Differentiable(sympy.Mul(*[self.expr, other], evaluate=False),
+                                  fd=self.fd)
+
+    __rmul__ = __mul__
+
+    def __imul__(self, other):
+        if isinstance(other, Differentiable):
+            self._expr = sympy.Mul(*[self.expr, other.expr], evaluate=False)
+        else:
+            self._expr = sympy.Mul(*[self.expr, other], evaluate=False)
+        return self
+
+    def __truediv__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(self.expr * sympy.Pow(other.expr, -1, evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(self.expr * sympy.Pow(other, -1, evaluate=False),
+                                  fd=self.fd)
+
+    def __rtruediv__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(other.expr * sympy.Pow(self.expr, -1, evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(other * sympy.Pow(self.expr, -1, evaluate=False),
+                                  fd=self.fd)
+
+    def __neg__(self):
+        return -1.0 * self
+
+    def __pow__(self, exponent):
+        return Differentiable(sympy.Pow(self, exponent, evaluate=False), fd=self.fd)
+
+    def __eq__(self, other):
+        if isinstance(other, Differentiable):
+            return self.expr.args == other.expr.args
+        elif isinstance(other, sympy.Expr):
+            return self.expr.args == other.args
+        else:
+            return sympy.simplify(self.expr) == other
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def func(self, *args, **kwargs):
+        kwargs.pop("evaluate", None)
+        if hasattr(self.expr, 'base'):
+            return self
+        else:
+            return Differentiable(self.expr.func(*args, **kwargs), fd=self.fd)
+
+    def __str__(self):
+        return self.expr.__str__()
+
+    __repr__ = __str__
+
+    @property
+    def args(self):
+        if hasattr(self.expr, 'base'):
+            return (self.expr,)
+        return self.expr.args
+
+    @property
+    def is_TimeFunction(self):
+        """
+        Check wether or not time is an index
+        """
+        return any(i.is_Time for i in self.indices)
+
+    def _space_order(self):
+        """
+        Infer space_order from expression
+        """
+        func = list(retrieve_functions(self.expr))
+        order = 100
+        for i in func:
+            order = min(order, getattr(i, 'space_order', order))
+
+        return order
+
+    def _time_order(self):
+        """
+        Infer space_order from expression
+        """
+        func = list(retrieve_functions(self.expr))
+        order = 100
+        for i in func:
+            order = min(order, getattr(i, 'time_order', order))
+
+        return order
+
+    def _dtype(self):
```

Comment:
_infer_dtype

Reply:
also, the implementation of this method should be revisited. The Functions have a well-defined dtype, why assuming it be either float or double? how about complexes or float16? 
is this any useful: https://github.com/opesci/devito/blob/master/devito/ir/clusters/cluster.py#L87

Reply:
Needs robustness added, smae for staggered that is incorrect for now. Will look at it